web/server/go.md

```diff
@@ -0,0 +1,498 @@
+  keeps code changes from propagating across the codebase
+  and becoming unmaintainable.
+
+#### Testing Models
+
+Theoretically models should have little responsibility
+other than struct definition.
+Any exported methods should be unit tested,
+such as data validations
+or state change methods, like above.
+
+#### Model References
+
+* [SQL Standard Library](https://golang.org/pkg/database/sql/)
+  and some [helpful examples](https://golang.org/src/database/sql/example_test.go)
+* [Illustrated Guide to SQLX](https://jmoiron.github.io/sqlx/)
+
+### Services
+
+So, we have handlers,
+which delegate a request/reponse flow
+and models that describe data.
+Where does the application logic go?
+Into the service layer!
+The service layer is where work gets delegated from a handler.
+The pattern is borrowed from other languages
+[like Ruby](https://medium.com/selleo/essential-rubyonrails-patterns-part-1-service-objects-1af9f9573ca1),
+where controllers and models
+get encumbered with multiple responsibilities.
+The service layer becomes a space
+for single use, reusable objects (service objects)
+to be built out.
+This layer will often be referred to as "business logic"
+or "application logic",
+but it's more or less
+what makes your application/API useful to a client.
+It's also where the various package dependencies
+of the application
+are likely to meet,
+answering the question of
+"how do I accomplish the client need with the given dependencies?"
+
+Let's take a look at an example!
+In our dog application,
+it's unlikely all we want to do
+is create dogs in a database.
+Let's say the purpose of the application
+is to show the user how cute a dog is
+given various attributes.
+
+Our handler may look something like this:
+
+```go
+package handlers
+
+type CalculateCutenessScore = func(dogID uuid.UUID) (int, error)
+
+type getCutenessScoreHandler struct {
+    calculateCutenessScore CalculateCutenessScore
+    //logger implementation may vary, we'll psuedocode for this example
+    logger Logger
+}
+
+func (h *getCutenessScoreHandler) handleGetCutenessScore(w http.ResponseWriter, r *http.Request) {
+    // like above we've deferred marshalling for this example
+    p := dogGetParams(w)
+
+    // offload calculation to service function
+    score, err := h.calculateCutenessScore(p.DogID)
+
+    if err == errors.CatchableError {
+        h.logger.Error(fmt.Sprintf("Could not calculate cuteness for dog %v", p.DogID), err)
+        w.Write(cutnessPayload(DEFAULT_CUTENESS_SCORE))
+    } else if err != nil {
+        http.Error(w, "Could not calculate dog's cuteness!", http.StatusInternalServerError)
+        return
+    }
+
+    w.Write(cutenessPayload(score))
+}
+```
+
+The handler here is not too much different than our `GetDog` example above.
+We've changed from `GetDog` to `GetCutenessScore`
+and likewise updated the service definition.
+We've also updated the error catching to be a little more robust.
+In this case,
```

Comment:
I'm hoping this doesn't cause confusion. I want to emphasize that the interface between handlers and services is a great way to dissect errors, rather than just pass whatever you get with a 500 attached.